Ft/versionable tables resource fix

Gustascct opened the pull request:
This PR does:

1. Fixes a bug where importing DSA with no resource automatically created a Saugyklos API distribution and assigned it to models.
2. Implements versioning for resource pages and resources themselves. This is done in this manner:
- - Creating a DatasetDistribution **does not** assign it a metadata_version and does not create a metadata instance.
- - When a DatasetDistribution is referenced by a model, then it **gets** the metadata_version and its metadata instance.
- - When all models stop referencing a DatasetDistribution, its metadata instance gets **deleted** and its metadata_version gets **removed**.
3. Creates two migrations:
- 1. First one creates a version for all datasets that do not have one yet.
- 2. The second one creates a Dataset metadata instance for datasets which do not have one yet. It also fixes DatasetDistributions, by following the rules outlined in 1 and 2.
4. Updates the DatasetDistribution display window. Adds a version picker which lets you choose the version and shows the DatasetDistributions which have the same version **or** the DatasetDistributions that do not have a version.

**Workaround for Metadata name:** The name generating function does not enforce the limit of 255, so to make the migration work I have increased the limit to 400. I checked out the Datasets which were hitting this limit on the test environment, I could not produce the error that I can produce it on my data dump.

Tests coming in another PR

Gustascct committed:
Added version fields

Gustascct committed:
Disabled translations

# Conflicts:
#	scripts/geoportal_import.py
#	vitrina/datasets/models.py
#	vitrina/resources/models.py

Gustascct committed:
Changed the names from Version to Publish

Gustascct committed:
Imported metadata gets a version

Gustascct committed:
Version should be selectable

Gustascct committed:
Version fields in tables

Gustascct committed:
Imported metadata gets a version immediately

Gustascct committed:
Main url changes

Gustascct committed:
View changes to allow versions

Gustascct committed:
FE supports new urls

Gustascct committed:
Url comments for Gustas

Gustascct committed:
Version field population of old data

Gustascct committed:
Unnecesary urls made versionable

Gustascct committed:
Temporary dropdown menu

Gustascct committed:
Version picker not temporary

Gustascct committed:
Made Base urls versionable

Gustascct committed:
added defaualt version

Gustascct committed:
changed the order of functions

Gustascct committed:
Fixed a few tests

Gustascct committed:
Reverted functions, added version deletes on error

Gustascct committed:
Added versioning to model-data url

Gustascct committed:
Fixed an error when structure errored but version was still accesible

Gustascct committed:
Supporting a few more urls with versioning

Gustascct committed:
Test changes part 1

Gustascct committed:
Fixed version-list and detail views

Gustascct committed:
Rewrote all tests that do not require publish version

Gustascct committed:
Revert "Fixed a few tests"

This reverts commit f2b80b1eb040fa39d06c640c1055e6d9de8badd2.

Gustascct committed:
Revert "Rewrote all tests that do not require publish version"

This reverts commit 32e9d65917f1db57ffd3afcc930ac8d935baa129.

Gustascct committed:
Revert "Test changes part 1"

This reverts commit 8daf5815e3a5c5cdcf418263a300a8e3a81072c1.

Gustascct committed:
Renamed variables

Gustascct committed:
Revert "Disabled translations"

This reverts commit 78f545039e32640d1dc60536e0252a2f9d478d08.

Gustascct committed:
Translations enabled and linter

Gustascct committed:
Handle importing DSA on the wrong version

Handle importing DSA on the wrong version

Gustascct committed:
Handle problem when version is None

Gustascct committed:
Reverse migration change

Gustascct committed:
Functions for field duplication upon publishing

# Conflicts:
#	vitrina/structure/views.py

Gustascct committed:
Metadata dataset gets a version

Gustascct committed:
Did a merge

Gustascct committed:
Reapply "Fixed a few tests"

This reverts commit bfcfe6280ea74bb4958c9428c7bb3e1d2f1135c8.

Gustascct committed:
Revert tests

Gustascct committed:
Reapply "Rewrote all tests that do not require publish version"

This reverts commit 6a98e863d9a4dc582e77598600e39029ed3303f1.

Gustascct committed:
Rewrote the files to use metadata_version instead of version

Gustascct committed:
Fixed test factories

Gustascct committed:
Fixed migration dependency issues

Gustascct committed:
Added filtering by version in services

Gustascct committed:
Creates PropertyLists with version

Gustascct committed:
Fixed a bug

Gustascct committed:
linter

Gustascct committed:
Craete duplicating fields function:

Gustascct committed:
Remove comments

Gustascct committed:
Renamed variables

Gustascct committed:
Changed variable to pk

Gustascct committed:
Fixed mistake

Gustascct committed:
temporary fix for tests

Gustascct committed:
bug fixes

Gustascct committed:
Fixed a few mistakes

Gustascct committed:
Random fixes

Gustascct committed:
Test changes

Gustascct committed:
Refactor, fix a few urls

Gustascct committed:
Fixed a mistake

Gustascct committed:
Merge commit

Gustascct committed:
Creates metadataversion now

Gustascct committed:
Fixed breadcrumbs

Gustascct committed:
Make migration better

Gustascct committed:
translations

Gustascct committed:
Variable names change

Gustascct committed:
moved version picker

Gustascct committed:
Added unnecesary version pickers, commiting if need to revert

Gustascct committed:
Revert "Added unnecesary version pickers, commiting if need to revert"

This reverts commit a898f64a56fca87ea394e5ae4bf50c88b5d275de.

Gustascct committed:
added version picker for dataset-structure-history only

Gustascct committed:
Review changes part 1

Gustascct committed:
Review changes part 2

Gustascct committed:
Linter

Gustascct committed:
Did a merge on ft/versioning-tables-version-picker

Gustascct committed:
fixed a bug

Gustascct committed:
Adds version number for versions that do not have one

Gustascct committed:
Field duplication WIP

Gustascct committed:
Generic version picker

Gustascct committed:
Redirect works if no version selected

Gustascct committed:
Make sure fields do not get created when fail

Gustascct committed:
A few breadcrumb fixes

Gustascct committed:
A few missed bugs

Gustascct committed:
Test rewrite

Gustascct committed:
Duplication logic (subject to change)

Gustascct committed:
minor bug fixes and formating

Gustascct committed:
Tests and translations

Gustascct committed:
Merge

Gustascct committed:
minor test fixes

Gustascct committed:
A few more tests

Gustascct committed:
Merge branch 'devel' into ft/versionable-tables-version-picker

Gustascct committed:
Merge branch 'ft/versionable-tables-version-picker' into ft/versionable-tables-duplicates-fields

Gustascct committed:
Review changes

Gustascct committed:
Revert "Review changes"

This reverts commit 91243b46fd269ef02c012bc192499d5d289afd27.

Gustascct committed:
Reapply "Review changes"

This reverts commit 9f765187adbeefd8363db8dde8cb5f06826759a8.

Gustascct committed:
Changed tests according to changes

Gustascct committed:
Minor logic changes

Gustascct committed:
Made model update not dependant on name

Gustascct committed:
Review changes and translations

Gustascct committed:
Type -> isinstance

Gustascct committed:
Custom typehints

Gustascct committed:
Format

Gustascct committed:
Translations and moved from init

Gustascct committed:
Refactor main function

Gustascct committed:
format

Gustascct committed:
Enum fix

Gustascct committed:
Partial fix for resources

Gustascct committed:
Migration for fixing data

Gustascct committed:
Fixed tests and logic

Gustascct committed:
Formatted

Gustascct committed:
Comment fix

Gustascct committed:
Removed a check for input resource

Gustascct committed:
format

Gustascct committed:
Fixed button alignment

Gustascct committed:
Revert "Fixed button alignment"

This reverts commit 1dbe079a1c6b1ec8833d1a385c6b95183e5439a3.

Gustascct committed:
Fixed buttons

Gustascct committed:
Did a merge on devel

Gustascct committed:
Did a merge on ft/versionable-tables-version-picker

Gustascct committed:
formatted

Gustascct committed:
Merge branch 'ft/versionable-tables-duplicates-fields' into ft/versionable-tables-resource-fix

Gustascct committed:
Made resources not create excess resources

Gustascct committed:
Additional migration, minor fixes

Gustascct committed:
Commit for cherrypick

Gustascct committed:
Workaround found

Gustascct committed:
Revert "Workaround found"

This reverts commit a644c7737c26ace25a436e13b6f05f46f41471cb.

Gustascct committed:
Reapply "Workaround found"

This reverts commit 04c83ebad0445c10365428483e3c8d6829b5bb26.

Gustascct committed:
Moved enum to init, added new functions

Gustascct committed:
Main changes to the resources

Gustascct committed:
Partial migration rewrite

Gustascct committed:
Migration finished, additional logic

Gustascct committed:
Minor fixes, migration squash

Gustascct committed:
restored test file changes

Gustascct committed:
Populate new fields of DD

Gustascct committed:
Name length increase

Gustascct committed:
Temp

Gustascct committed:
Minor changes

martynasracct commented on vitrina/datasets/services.py:
```diff
@@ -432,9 +435,9 @@ def _generate_download_url(self, models, distribution_format):
         return download_url, distribution_name
 
-    def retrieve_data(self, dataset_pk, resource_name, distribution_format):
+    def retrieve_data(self, dataset_pk, metadata_version, resource_name, distribution_format):
```
typing-hints missing

Gustascct replied:
Added

martynasracct commented on vitrina/datasets/templates/vitrina/datasets/detail.html:
```diff
@@ -158,11 +158,14 @@
             <div class="column is-full">
                 <h2 class="custom-title is-size-4-mobile no-margin-bottom">{% translate "Duomenų šaltiniai" %}</h2>
-                {% include 'vitrina/datasets/datatable.html' %}
+                {% include 'vitrina/datasets/datatable.html'%}
```
```suggestion
                {% include 'vitrina/datasets/datatable.html' %}
```

Gustascct replied:
changed

martynasracct commented on vitrina/datasets/views.py:
```diff
@@ -459,6 +459,21 @@ def has_permission(self):
+        if not metadata_version:
+            metadata_version = _Version.objects.filter(dataset=self.object).order_by("version").last()
+
+            if metadata_version:
```
```suggestion
            if metadata_version := _Version.objects.filter(dataset=self.object).order_by("version").last()
```

Gustascct replied:
changed

martynasracct commented on vitrina/datasets/views.py:
```diff
@@ -459,6 +459,21 @@ def has_permission(self):
+    def get(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        metadata_version = request.GET.get("resource_version")
+        if not metadata_version:
```
```suggestion
        if not metadata_version := request.GET.get("resource_version")
```

Gustascct replied:
Changed

martynasracct commented on vitrina/datasets/views.py:
```diff
@@ -459,6 +459,21 @@ def has_permission(self):
+
+            if metadata_version:
+                url = (
+                    reverse("dataset-detail", kwargs={"pk": self.object.pk})
+                    + f"?resource_version={metadata_version.pk}"
+                )
```
```suggestion
                url = f'{reverse("dataset-detail", kwargs={"pk": self.object.pk})}?resource_version={metadata_version.pk}'
```

Gustascct replied:
changed

martynasracct commented on vitrina/datasets/views.py:
```diff
@@ -470,6 +485,13 @@ def get_queryset(self) -> QuerySet[Dataset]:
         context_data = super().get_context_data(**kwargs)
         dataset = context_data.get("dataset")
+        dataset_versions = _Version.objects.filter(dataset=dataset).order_by("version")
+        context_data["versions"] = dataset_versions
```
```suggestion
        context_data["versions"] = _Version.objects.filter(dataset=dataset).order_by("version")
```

Gustascct replied:
Changed

martynasracct commented on vitrina/datasets/views.py:
```diff
@@ -470,6 +485,13 @@ def get_queryset(self) -> QuerySet[Dataset]:
+        context_data["versions"] = dataset_versions
+        metadata_version_id = self.request.GET.get("resource_version")
+        metadata_version = None
+        if metadata_version_id:
```
```suggestion
        metadata_version = None
        if metadata_version_id := self.request.GET.get("resource_version")
```

Gustascct replied:
changed

martynasracct commented on vitrina/resources/models.py:
```diff
@@ -129,7 +130,7 @@ class DatasetDistribution(TranslatableModel):
     deleted = models.BooleanField(blank=True, null=True)
     deleted_on = models.DateTimeField(blank=True, null=True)
-    dataset = models.ForeignKey(Dataset, models.CASCADE)
+    dataset = models.ForeignKey("vitrina_datasets.Dataset", models.CASCADE)
```
This change was needed because of circular import?

Gustascct replied:
Yes it was

martynasracct commented on vitrina/resources/models.py:
```diff
@@ -342,7 +354,18 @@ def get_acl_parents(self):
         return parents
 
     def get_absolute_url(self):
```
```suggestion
    def get_absolute_url(self) -> str:
```

Gustascct replied:
changed

martynasracct commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+                    .order_by("name")
+                    .values_list("name", flat=True)
+                    .last()
+                )
```
`name__iregex=r"resource[0-9]+",` looks scary. Why only `Metadata` objects with these symbols specifically? What other `Metadata` names can be there?

Gustascct replied:
I took this entire code from another place, so it was not written by me. Should I rewrite it?

martynasracct replied:
Makes sense. I think you can keep it in that case 🤷‍♂️

martynasracct commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+        else:
+            name = self.name
+            if not name:
+                name = (
```
```suggestion
                latest_metadata_name = (
```

Gustascct replied:
changed

martynasracct commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+                    except ValueError:
+                        n = 0
+                    n += 1
+                    name = f"resource{n}"
```
This part is confusing, I suggest moving it into separate function. Function should make this code more descriptive. For example:

```python
def _get_next_metadata_name(previous_metadata_name: str | None) -> str:
    """
    Get version from previous_metadata_name and increase it by 1. Return 1 if previous_metadata_name is None
    """
    if not previous_metadata_name:
        previous_metadata_name = "resource0"

    version_string = previous_metadata_name.replace("resource", "")
    version_number = int(version_string) if version_string.isdigit() else 0
    
    return f"resource{version_number + 1}"
```

It's just an example, maybe you will think of a better way to do this

Gustascct replied:
This part was taken from a previous implementation as discussed before so left it as is. Improved the naming

martynasracct commented on vitrina/resources/urls.py:
```diff
@@ -11,20 +11,40 @@
+        "resource/<int:pk>/change",
+        ResourceUpdateView.as_view(),
+        name="resource-change-no-version",
+    ),
```
This URL should fit into single line

Gustascct replied:
Changed

martynasracct commented on vitrina/resources/views.py:
```diff
@@ -104,42 +102,6 @@ def form_valid(self, form):
         resource.dataset = self.dataset
         resource.save()
 
```
Just checking: you moved this code into `create_metadata_instance_and_assign_version` method. Don't you need to call this method here?

Gustascct replied:
No, the logic changed. Now a resource is created unversioned and without a metadata row. Only after it is referenced the version and metadata get created

martynasracct commented on vitrina/resources/views.py:
```diff
@@ -343,10 +306,11 @@ def has_permission(self):
         return has_perm(self.request.user, Action.VIEW, dataset)
 
-    def get_data(self, dataset_pk, model_name, distribution_format):
+    def get_data(self, dataset_pk, metadata_version_pk, model_name, distribution_format):
```
type checking missing

Gustascct replied:
changed

martynasracct commented on vitrina/structure/migrations/0012_populate_version_of_new_tables.py:
Are these migrations already merged to main branch?

Gustascct replied:
Not merged on devel

martynasracct replied:
All good then

martynasracct commented on vitrina/structure/migrations/0013_version_number_assigning.py:
Are these migrations already merged to main branch?

Gustascct replied:
Not merged on devel

martynasracct replied:
All good then

martynasracct commented on vitrina/structure/models.py:
```diff
@@ -304,14 +303,45 @@ def get_base_props(self):
+            old_distribution = old_instance.distribution
+
+        if self.distribution:
+            if self.distribution.format.extension == "UAPI":
```
```suggestion
        if self.distribution and self.distribution.format.extension == "UAPI":
```

Gustascct replied:
Changed

martynasracct commented on vitrina/resources/migrations/0012_datasetdistribution_access_datasetdistribution_level_and_more.py:
```diff
@@ -0,0 +1,44 @@
+
+    for dataset_distribution in DatasetDistribution.objects.all():
+        metadata_instance = Metadata.objects.filter(content_type=dd_ct, object_id=dataset_distribution.id).first()
+        if metadata_instance:
```
```suggestion
        if metadata_instance := Metadata.objects.filter(content_type=dd_ct, object_id=dataset_distribution.id).first():
```

Gustascct replied:
Changed

martynasracct commented on vitrina/structure/migrations/0011_create_version_for_datasets_with_no_version.py:
```diff
@@ -0,0 +1,29 @@
+    Version = apps.get_model('vitrina_structure', 'Version')
+
+    for dataset in Dataset.objects.all():
+        if not dataset.dataset_version.exists():
```
```suggestion
    for dataset in Dataset.objects.filter(dataset_version__isnull=True).prefetch_related("metadata_set"):

```

Gustascct replied:
Changed

martynasracct commented on vitrina/structure/migrations/0013_dataset_datasetdistribution_data_fix.py:
Scary migration 😨

Gustascct replied:
Yes, but the logic was approved by Mantas

Gustascct committed:
PR fixes

Gustascct committed:
Additional check

Struchid commented on vitrina/datasets/templates/vitrina/datasets/detail.html:
```diff
@@ -162,7 +162,10 @@ <h2 class="custom-title is-size-4-mobile no-margin-bottom">{% translate "Duomen
-        {% include 'vitrina/datasets/add_resource.html' %}
+        <div class="buttons is-right m-t-md">
+            {% include "version_picker.html" with version_as_query_param=True url_name="dataset-detail" dataset=dataset versions=versions selected_version=selected_version %}
+            {% include 'vitrina/datasets/add_resource.html' %}
```
Let's keep consistent


```suggestion
            {% include "vitrina/datasets/add_resource.html" %}
```

Gustascct replied:
Changed

Struchid commented on vitrina/resources/migrations/0012_datasetdistribution_access_datasetdistribution_level_and_more.py:
```diff
@@ -0,0 +1,43 @@
+    dd_ct = ContentType.objects.get_for_model(DatasetDistribution)
+
+    for dataset_distribution in DatasetDistribution.objects.all():
+        if metadata_instance := Metadata.objects.filter(content_type=dd_ct, object_id=dataset_distribution.id).first():
```
This is the n+1 problem, even though these are migrations, might be worth improving on it

Gustascct replied:
Fetched all metadata rows for datasetdistributions, created a dictionary out of them and accessed values as needed. By my understanding it should make it better

Struchid commented on vitrina/resources/models.py:
```diff
@@ -341,8 +353,19 @@ def get_acl_parents(self):
-    def get_absolute_url(self):
-        return reverse("resource-detail", kwargs={"pk": self.dataset.pk, "resource_id": self.pk})
+    def get_absolute_url(self) -> str:
+        if self.metadata_version and self.metadata_version.pk:
```
Is there a possibility that `metadata_version` will be unsaved for the `DatasetDistribution`, or is this (`self.metadata_version.pk`) more of a sanity check?

Gustascct replied:
Yes there is, because an unasigned DatasetDistribution does not have Metadata, therefore no version.

Struchid commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
             entry.update_description()
+
+    def create_metadata_instance_and_assign_version(self, metadata_version: int) -> Metadata:
+        metadata_version = Version.objects.get(pk=metadata_version)
```
Let's add some safeguards here, if the Version you would try to `get()` does not exist, otherwise this will fail hard for the user.

Let's add logs for that too

Gustascct replied:
added

Struchid commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+                if not latest_metadata_name:
+                    name = "resource1"
+                else:
+                    n = latest_metadata_name.replace("resource", "")
```
Let's not use one letter variables (if we're already touching that part of the code), e.g.:

```suggestion
                    duplicate_name_suffix = latest_metadata_name.replace("resource", "")
```

Gustascct replied:
Added

Struchid commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+            metadata_instance = Metadata.objects.create(
+                uuid=str(uuid.uuid4()),
+                dataset=self.dataset,
+                content_type=ContentType.objects.get_for_model(DatasetDistribution),
```
Called twice, could be called once at the beginning of the method

Gustascct replied:
Removed

Struchid commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+                    n += 1
+                    name = f"resource{n}"
+                self.name = name
+            self.save()
```
Seems like `update_fields` could be used, due to only updating one attribute

Gustascct replied:
This part was removed

Struchid commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,56 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
         for entry in all_entries.filter(url__in=new_urls):
             entry.update_description()
+
+    def create_metadata_instance_and_assign_version(self, metadata_version: int) -> Metadata:
```
The method seems to not necessarily create new metadata instance, but might re-use an older instance

Gustascct replied:
renamed

Struchid commented on vitrina/resources/views.py:
```diff
@@ -370,9 +335,13 @@ def get(self, request, *args, **kwargs):
+            else None,
+            "api_url": reverse("getall-api", args=[self.dataset.pk, metadata_version_pk, self.models[0].name])
+            if self.models
+            else None,
```
I'd extract the `self.models[0].name` above to avoid duplication

Gustascct replied:
extracted

Struchid commented on vitrina/structure/migrations/0011_create_version_for_datasets_with_no_version.py:
```diff
@@ -0,0 +1,28 @@
+            created_draft_version.save()
+            for metadata in dataset.metadata_set.all():
+                metadata.metadata_version = created_draft_version
+                metadata.save()
```
Seems like this is indented 2x, unless it is github with its formatting nonsenses again

Gustascct replied:
yep good catch, unindented

Struchid commented on vitrina/structure/migrations/0011_create_version_for_datasets_with_no_version.py:
```diff
@@ -0,0 +1,28 @@
+
+    for dataset in Dataset.objects.filter(dataset_version__isnull=True).prefetch_related("metadata_set"):
+            created_draft_version = Version.objects.create(dataset=dataset, version=1, status=VersionStatus.DRAFT)
+            created_draft_version.save()
```
You are already doing create and not just instantiating an instance of the Version model, no need to do the save once more.

Gustascct replied:
Removed

Struchid commented on vitrina/structure/forms.py:
```diff
@@ -463,6 +464,12 @@ class ModelCreateForm(forms.ModelForm):
         help_text=_("Savybė nurodanti modelio metaduomenų gyvavimo ciklo būseną."),
     )
+    distribution = forms.ModelChoiceField(
+        label=_("Duomenų distribucija"),
```
IMO It should translate to `pateiktis` instead of `distribucija`, but I might be missing some context and therefore lying

Gustascct replied:
Yeah not sure, but when I presented it there were no objections. Will change if people start objecting if thats okay with you

Struchid replied:
Agreed, let's leave it as is then

Struchid commented on vitrina/structure/models.py:
```diff
@@ -304,14 +303,44 @@ def get_base_props(self):
     def is_opened(self):
         return self.dataset.is_opened()
 
+    def save(self, *args, **kwargs):
```
Let's add return typehints for the methods in this file

Gustascct replied:
added

Struchid commented on vitrina/structure/views.py:
```diff
@@ -3647,6 +3654,30 @@ def create_related_model_duplicate(self, old_metadata_instance: Metadata) -> tup
         new_related_instance.save()
 
+        if isinstance(old_related_instance, DatasetDistribution):
+            for trans in old_related_instance.translations.all():
```
Let's use full names instead, wouldn't want to cause any confusion what we're looping through 😶

```suggestion
            for translation in old_related_instance.translations.all():
```

Gustascct replied:
changed

wisecrow commented:
Kai bandau importuoti naują struktūrą https://github.com/atviriduomenys/demo-saltiniai/blob/main/manifest/datasets/gov/vssa/demo/wsdl/wsdl_rctest.csv
gaunu klaidą- 
```
File "/app/vitrina/structure/services.py", line 155, in _load_datasets
_link_distributions(meta, dataset, metadata_version)
File "/app/vitrina/structure/services.py", line 636, in _link_distributions
model.save()
File "/app/vitrina/structure/models.py", line 321, in save
if self.distribution and self.distribution.format.extension == "UAPI":
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'extension'
```

Gustascct committed:
Simplified name assigning

Gustascct committed:
Formatted and enabled rdf download which download all version distributions

gts446 commented on vitrina/datasets/services.py:
```diff
@@ -321,19 +322,22 @@ class DynamicResourceService:
+        if is_for_rdf_export:
+            models = self.dataset.model_set.filter()
+        else:
+            models = self.dataset.model_set.filter(metadata_version=self.metadata_version)
```
```suggestion
        models = self.dataset.model_set.filter()
        if is_for_rdf_export:
            models = models.filter(metadata_version=self.metadata_version)
```

Struchid replied:
The suggestion is invalid and changes the flow. Also, why do we need the `filter()` with no attributes?

```suggestion
        models = self.dataset.model_set  # might need `.all()` here
        if not is_for_rdf_export:
            models = models.filter(metadata_version=self.metadata_version)
```

Gustascct replied:
Changed

gts446 commented on vitrina/datasets/views.py:
```diff
@@ -470,6 +482,11 @@ def get_queryset(self) -> QuerySet[Dataset]:
         dataset = context_data.get("dataset")
+        context_data["versions"] = _Version.objects.filter(dataset=dataset).order_by("version")
+        metadata_version = None
+        if metadata_version_id := self.request.GET.get("resource_version"):
```
Is it possible that this will be None?

gts446 commented on vitrina/datasets/views.py:
```diff
@@ -459,6 +459,18 @@ def has_permission(self):
+                url = (
+                    f"{reverse('dataset-detail', kwargs={'pk': self.object.pk})}?resource_version={metadata_version.pk}"
+                )
+                return redirect(url)
```
Might wanna add `else` here to validate that version actually exists here, otherwise user can pass anything here and later parts will fail.

gts446 commented on vitrina/datasets/views.py:
```diff
@@ -459,6 +459,18 @@ def has_permission(self):
+    def get(self, request, *args, **kwargs):
+        self.object = self.get_object()
+
+        if not request.GET.get("resource_version"):
+            if metadata_version := _Version.objects.filter(dataset=self.object).order_by("version").last():
+                url = (
+                    f"{reverse('dataset-detail', kwargs={'pk': self.object.pk})}?resource_version={metadata_version.pk}"
+                )
+                return redirect(url)
```
```suggestion
        if resource_version := request.GET.get("resource_version"):
            try:
                self.metadata_version = _Version.objects.get(pk=resource_version)
                return super().get(request, *args, **kwargs)
            except:
                pass
       
        if metadata_version := _Version.objects.filter(dataset=self.object).order_by("version").last():
            url = (
                f"{reverse('dataset-detail', kwargs={'pk': self.object.pk})}?resource_version={metadata_version.pk}"
            )
            return redirect(url)
        self.metadata_version = None
```

Also you can remove later code in this view that retrieves `metadata_version` and use `self.metadata_version`

Struchid replied:
What's the point of a bare except statement here? `get_or_404` will be more than enough.

> Also you can remove later code in this view that retrieves metadata_version and use self.metadata_version

Ambiguous comment, and why does this suggestion imply that setting `self.metadata_version` inside of a views GET method a good practice? It is a request handling method, not a state container and setting the self.metadata_version and using it elsewhere expecting it to be set is bound to bear bad news in the future.

I do not understand how is this suggestion better than what's already implemented, make sure to write an explanation to your suggestions

gts446 replied:
Depends wether we want to show 404 for non existing version or default to existing one. I think defaulting is better, but I guess thats opinionated.
Regarding `self.metadata_version` I see no point in additional querying of `metadata_version`, because its not going to change throughout the view. Save it to the view on first occurence if its being used later on in view, or save it to variable if its not. I don't understand why thats not a good practice?

Gustascct replied:
The metadata_version comes from a GET request so it is safer to keep it as a parameter only on the same request. The get method only handles redirecting, get_context_data is the only function that uses metadata_version so there is no reason to make them depend on each other.

gts446 commented on vitrina/datasets/views.py:
```diff
@@ -470,6 +482,11 @@ def get_queryset(self) -> QuerySet[Dataset]:
         context_data = super().get_context_data(**kwargs)
         dataset = context_data.get("dataset")
+        context_data["versions"] = _Version.objects.filter(dataset=dataset).order_by("version")
+        metadata_version = None
+        if metadata_version_id := self.request.GET.get("resource_version"):
+            metadata_version = _Version.objects.filter(id=metadata_version_id).first()
+            context_data["selected_version"] = metadata_version
```
```suggestion
        context_data["selected_version"] = self.metadata_version
```

Struchid replied:
If this suggestion comes from these assumptions:
- https://github.com/atviriduomenys/katalogas/pull/2233#discussion_r2697379756

Nope.

gts446 commented on vitrina/datasets/views.py:
```diff
@@ -511,8 +527,19 @@ def get_context_data(self, **kwargs):
             "licences": set([dist.licence for dist in dataset.datasetdistribution_set.filter(licence__isnull=False)]),
         }
 
+        if metadata_version.status == VersionStatus.DRAFT:
```
This will crash if metadata_version is None.

Also update instances of `metadata_version` to `self.metadata_version`

Gustascct replied:
Added the check, metadata_version is contained in a singular function and it comes from a GET request, so its safer to keep it only on the request

gts446 commented on vitrina/resources/migrations/0012_datasetdistribution_access_datasetdistribution_level_and_more.py:
```diff
@@ -0,0 +1,43 @@
+            dataset_distribution.name = metadata_instance.name
+            dataset_distribution.access = metadata_instance.access
+            dataset_distribution.level = metadata_instance.level
+            dataset_distribution.save()
```
As I understand here `Metadata` can be linked with `DatasetDistribution` through `GenericForeighKey`? In that case it would make sense to add new virtual field `metadata = GeneriRelation('Metadata')` to `DatasetDistribution`. and then change code to 
```suggestion
    for dataset_distribution in DatasetDistribution.objects.filter(metadata__isnull=False).select_related("metadata"):
        if metadata_instance := dataset_distribution.metadata.first():
            dataset_distribution.name = metadata_instance.name
            dataset_distribution.access = metadata_instance.access
            dataset_distribution.level = metadata_instance.level
            dataset_distribution.save()
```

Gustascct replied:
DatasetDistribution does have this field already, but select_related is incorrect and prefetch_related works in shell but does not work in the migration for some reason. Either way made a speed improvement

gts446 commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,32 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+
+    def create_metadata_instance_and_assign_version(self, metadata_version: int) -> Metadata:
+        metadata_version = Version.objects.get(pk=metadata_version)
+        if metadata_instance := self.metadata.first():
```
```suggestion
    def create_metadata_instance_and_assign_version(self, metadata_version_id: int) -> Metadata:
        metadata_version = Version.objects.get(pk=metadata_version_id)
        if metadata_instance := self.metadata.first():
```

Gustascct replied:
changed

gts446 commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,32 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+        metadata_version = Version.objects.get(pk=metadata_version)
+        if metadata_instance := self.metadata.first():
+            metadata_instance.metadata_version = metadata_version
+            metadata_instance.save()
```
```suggestion
            metadata_instance.save(update_fields=["metadata_version"])
```

Gustascct replied:
updated

gts446 commented on vitrina/resources/models.py:
```diff
@@ -400,3 +423,32 @@ def update_applicable_legislation(self, urls: list[str]) -> None:
+        if not self.model_set.exists():
+            self.metadata_version = None
+            self.save()
+            self.metadata.all().delete()
```
By the name of this method it seems that this checks something and returns True/False or raises excepton.
My suggestion would be to do this check `if not self.model_set.exists():` whereever you need this method. and then maybe rename and change this method to:

```suggestion
    def delete_metadata(self):
        self.metadata_version = None
        self.save()
        self.metadata.all().delete()
```

Or leave as is but only rename method to `delete_metadata_if_has_no_models`.

Gustascct replied:
renamed

gts446 commented on vitrina/resources/views.py:
```diff
@@ -280,16 +231,22 @@ def handle_no_permission(self):
             resource = get_object_or_404(DatasetDistribution, id=self.kwargs["pk"])
+            if resource.metadata_version:
+                url = reverse("dataset-detail", kwargs={"pk": resource.dataset.pk})
+                return HttpResponseRedirect(f"{url}?resource_version={resource.metadata_version.pk}")
```
I suggest to move `resource` from variable to instance property inside `has_permission` and then call it wherever needed. Becuase now it is fetched at least three times (in `has_permisison`, `handle_no_permission`, `form_valid`)

Gustascct replied:
has_permission should run first, but it is not guaranteed, so setting self.resource in it would be incorrect because if it does not run the view will fail. Solving this problem would require adding a few functions which would have to be called from all 3 functions which would defeat the purpose.

gts446 replied:
Could you give example when `has_permission` would not run first? I might be missing something.

Gustascct replied:
I do not know but I think it is reasonable to have such a doubt. 

In addition, has_permissions is not the right place to define self variables

gts446 replied:
dispatch chain is `<..>->dispatch()->has_permission()-><..>` (Thats because this view inherits `PermisisonRequiredMixin`) if user has permission then it goes to whatever http method is and remaining logic.
If not in `has_permission` then you can declare it inside `dispatch`, or `setup` (which is run before dispatch), up to you. But either way `get_object_or_404(DatasetDistribution, id=self.kwargs["pk"])` should not be called three times in a single view.

Struchid replied:
There is no case at the top of my head, where it would be a good idea to set instance attributes inside anything else than the init method of a class (`setup()` for `TemplateView` classes), we do not want to bring in that kind of burden in the case that something is run before the methods like your suggested `dispatch()` is, even marked on your example that somethings are run before `dispatch()`.

Anyhow, this is not a breaking change and a lot of this is built atop of old logic, so refactoring every single place while only wanting to add this kind of minimal improvement is a big burden to the developer (Gustas in this case) and the tasks would grow massively. To avoid this, we could add a ticket for improvements and let the PO decide on how much this kind of an improvement would be a priority.

@Gustascct ultimately it is for you to decide, if you want to move the resource to instance var and to the `setup()` method and improve this class or you're just building atop of this already re-use full class. If not - resolve the issue.

The last two things — regarding code reviews. Threads like these are slow and can not be found easily enough and it's better to discuss it synchronically (either by a call or a talk live). The other thing - it is not nice to block fellow developers with `Request for changes` just to bring in these kind of improvements, so let's avoid using that button in the future and use it only on changes, that would most definitely break production once deployed.

gts446 replied:
`__init__` is definitely not the right place to declare view attributes.

PO has already mentioned that there won't be separate tasks for refactoring/old code improvements, but that could/should be done together with other tasks where implementation "touches" the code that needs optimization/refactoring.

This change is 1 minute of coding, and it forms a good practice. So not sure why this discussion is where it is...

Regarding `Request for change` has already been discussed with @lescijuscct

gts446 commented on vitrina/resources/views.py:
```diff
@@ -280,16 +231,22 @@ def handle_no_permission(self):
             return redirect(settings.LOGIN_URL)
         else:
```
no need for `else` here

Gustascct replied:
removed

gts446 commented on vitrina/resources/views.py:
```diff
@@ -307,6 +264,11 @@ def form_valid(self, form: BaseForm) -> HttpResponse:
+
+        if version_id:
+            url = reverse("dataset-detail", kwargs={"pk": dataset.pk})
+            return HttpResponseRedirect(f"{url}?resource_version={version_id}")
```
```suggestion
        with create_revision():
            add_to_revision(resource)
            set_user(self.request.user)
            set_comment((f'Ištrintas duomenų šaltinis "{resource.lt_title()}".'))
            resource.delete()
            resource.metadata.all().delete()

            if not DatasetDistribution.objects.filter(dataset=dataset) and dataset.is_public:
                if dataset.plandataset_set.exists():
                    dataset.status = Dataset.PLANNED
                    comment_status = Comment.PLANNED
                else:
                    dataset.status = Dataset.INVENTORED
                    comment_status = Comment.INVENTORED

                Comment.objects.create(
                    content_type=ContentType.objects.get_for_model(dataset),
                    object_id=dataset.pk,
                    type=Comment.STATUS,
                    status=comment_status,
                    user=self.request.user,
                )
                dataset.save()

        if  version_id := self.kwargs.get("version_id"):
            url = reverse("dataset-detail", kwargs={"pk": dataset.pk})
            return HttpResponseRedirect(f"{url}?resource_version={version_id}")
```

Gustascct replied:
changed

gts446 commented on vitrina/structure/migrations/0011_create_version_for_datasets_with_no_version.py:
```diff
@@ -0,0 +1,28 @@
+    Dataset = apps.get_model('vitrina_datasets', 'Dataset')
+    Version = apps.get_model('vitrina_structure', 'Version')
+
+    for dataset in Dataset.objects.filter(dataset_version__isnull=True).prefetch_related("metadata_set"):
```
Isnt `Dataset.version` deprecated? Shouldn't you check by `dataset__metadata__metadata_version`?

Gustascct replied:
Theres nothing to suggest that it is deprecated, why do you think that this is a problem?

gts446 replied:
Sorry, I mixed up `Dataset.version` with `Dataset.dataset_version`

gts446 commented on vitrina/structure/models.py:
```diff
@@ -304,14 +303,42 @@ def get_base_props(self):
+            old_distribution = old_instance.distribution
+
+        if self.distribution and self.distribution.format and self.distribution.format.extension == "UAPI":
+            raise ValidationError(_("Negalima priskirti Saugyklos API distribucijos. Pasirinkite kitą distribuciją."))
```
This should be at the top of the method.

Gustascct replied:
moved

gts446 commented on vitrina/structure/models.py:
```diff
@@ -304,14 +303,42 @@ def get_base_props(self):
+
+        super().save(*args, **kwargs)
+
+        if self.distribution:
+            if self.distribution.metadata_version and self.distribution.metadata_version != self.metadata_version:
+                return
```
`DatasetDistribution` should not have duplicated `metadata_version` field which is directly linked with `DatasetDistribution.metadata.metadata_version`

Gustascct replied:
I agree that this is not ideal but having the connection to Version only through Metadata is also not the best idea. There should be a table between Dataset and Metadata. Wont touch now, but an improvement can be made

gts446 replied:
Another thing. Shouldn't there be some exception raised or something on line 328 instead of empty return? I understand if we run into such case then it means serious problems, because data is corrupt, no ?

Gustascct replied:
No, this happens when publishing a version. At one moment there is a DatasetDistribution which does not match the version of the model, this has to be let through here. There are validations higher up that make sure that this does not cause a problem

gts446 commented on vitrina/structure/models.py:
```diff
@@ -304,14 +303,42 @@ def get_base_props(self):
+            self.distribution.save(update_fields=["metadata_version"])
+
+        if old_distribution:
+            old_distribution.check_if_resource_should_be_versioned()
```
Looking at this whole method I hardly understand what are you trying to achieve here.
I can see you validate that Model is not assigned to Saugykla, but that should be handled in form validations not in model.save().
Also could you explain the rest of the logic?
Why you create metadata for distribution only when it is assigned to model?
If that is some business requirement then I think this should also be done inside view, not model.save()

Gustascct replied:
This method ensures that the saving of a model and creation of needed datasetdistribution metadata is correct.

There are 3 ways to create a model, through a form, through a DSA import and one by publishing a model (it creates a duplicate). Having all of the logic in the save method saves me from having to duplicate the code in different paths and from having to customize it to each way of creation.

Also updating the models datasetdistribution should also run these checks, which do run if the code is in the save method. So in all, at least 4 distinct cases for when this code should run. 

I can explain the logic in a call.

Yes, there is a business requirement for it being created this way.

gts446 replied:
Then maybe add extra method(s) that would implement this logic and call it explicitly where needed before `save`? For me whole logic inside `save` looks scary and confusing, and the fact that its inside `save` makes it even scarier :). Maybe a 3rd opinion would be good here. @lescijuscct ?

Gustascct replied:
Asked for input from Gytis, Martynas and Darius, they agreed that having it in the save makes sense

Gustascct committed:
PR fixed and migrations change

Gustascct committed:
Gytis review and format

Gustascct committed:
PR fixes

Gustascct committed:
Translations

Gustascct committed:
Did a merge on devel

Gustascct committed:
Renamed variable

gts446 reviewed (commented):
.

Gustascct committed:
Service fix, forms

Gustascct committed:
Formatted

MindaugasStanionis committed:
Metadata, resource forms/views hiding in not draft versions (#2285)

* 2233: Adding logic in dispatch to not let in resource, structure updates/deletes. Adding logic to disable buttons for creating/updating metadata elements

# Conflicts:
#	tests/structure/test_views.py
#	vitrina/datasets/templates/vitrina/datasets/detail.html

* Fixing ruff problem

* 2284: PR fixes

* 2284: PR fixes

* 2293: PR fixes

* 2293: PR fixes

* 2293: Fixing linting

* 2293: Adding node to drone.yml for deploying to test env

* 2285: Fixing buttons that are not disbled

Gustascct committed:
Ft/versionable tables tests fix (#2235)

* Fixed a part of tests

* Almost all tests fixed

* Partial fix

* Revert "Commit for cherrypick"

This reverts commit 38b14e2a7ff5e0dd7a02b890720e077a543b7235.

# Conflicts:
#	vitrina/datasets/views.py

* Tests

* Factory rewrite

* Fixed dataset tests, factory update

* Fixed dataset tests, changed factories again

* Fixed a few tests

* Updated and new resource tests

* Simplified name assigning

* Fixed a part of tests, services rewerted

* Fixed factory, test fixes

* Test fixes and minor fixes

* Fixed most of the tests, changed uuid factory and comment fix

* Ruff:

* Format, service fix

* Fixed services, imported models

* Test fixes, indentation fixes

* did a merge and test fixes

* PR changes

* PR fix

* Removed breakpoint

* Permission handling and test fixes

* Permission handling

MindaugasStanionis committed:
2233: Adding new translations

Gustascct committed:
Merge branch 'devel' into ft/versionable-tables-resource-fix

Gustascct committed:
Merged on devel

Gustascct committed:
Test fixes

Gustascct committed:
Fixed model delete endpoint, url, tests

MindaugasStanionis committed:
2233: Adding permission to not let delete model on not draft versions

Gustascct committed:
Fixed comments form temporarily

Gustascct committed:
Renamed function, changed check

Gustascct committed:
Changes.rst

Gustascct committed:
Changes rst

Gustascct committed:
Added the possibility to disable publish button (#2313)

* Added the possibility to disable publish button

* Revert "Added the possibility to disable publish button"

This reverts commit e8601c9074c14e858738e00c7d29d17de5ed0ef2.

* Introduced django-flags and made a flag

* Required removed as it doe snothing with single conditions

* Format

* Changed default to False, removed the url altogether

Gustascct committed:
Translations

Gustascct committed:
Fuzzy translations

Gustascct committed:
Revert "Fuzzy translations"

This reverts commit 708e261b67796d235f33da090c404f8ad0ac2705.

Gustascct committed:
Revert "Translations"

This reverts commit 28f1edeead882aac9bda1a5773a742b430dadb50.
